keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1108 @@
+    // Deployments, CRDS, etc). The keys of the Components map are the identifiers of the components. These may not 
+    // be the names of the components. For example, two instances of an Application of the same type may use components 
+    // with same identifier key, but the names of the components must differ in order to ensure uniqueness in the 
+    // namespace. The values in the Components map are corresponding GroupKinds of the Application's components. These 
+    // may be used, in conjunction with the Selector, to retrieve an Application's components at the client's desired 
+    // version.
+    Components map [string]metav1.GroupKind `json:"components,omitempty" protobuf:"bytes,2,opt,name=components"`
+    
+    // Dependencies is a list of other Applications that this Application depends on.
+    Dependencies []sting `json:"dependencies" protobuf:"bytes,3,opt,name=dependencies"`
+    
+    // Selector is a label query over kinds that created by the application. It must match the component objects' labels.
+    // More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+    Selector *metav1.LabelSelector `json:"selector" protobuf:"bytes,4,opt,name=selector"`
+    
+    // HealthCheck is the name of an optional Deployment that provides the health status of the Application. When the 
+    // Deployment is healthy, the application is considered to be healthy.
+    HealthCheckDeployment string `json:"healthCheck" protobuf:"bytes,5,opt,name=healthCheck"`
+    
+    // Version is an optional version indicator for the application.
+    Version string `json:"version,omitempty" protobuf:"bytes,6,opt,name=version"`
+    
+    // Description is a brief string description of the application. 
+    Description string `json:"description,omitempty" protobuf:"bytes,7,opt,name=description"`
+    
+    // Maintainers is an optional list of maintainers of the application. The maintainers in this list are maintain the 
+    // the source code, images, and package for the application.
+    Maintainers []string `json:"maintainers,omitempty" protobuf:"bytes,8,opt,name=maintainers"`
+    
+    // Owners is an optional list of the owners of the installed application. The owners of the application should be 
+    // contacted in the event of a planned or unplanned disruption affecting the application.
+    Owners [] string `json:"owners,omitempty" protobuf:"bytes,9,opt,name=owners"`
+    
+    // Keywords is an optional list of key words associated with the application (e.g. MySQL, RDBMS, database).
+    Keywords []string `json:"keywords,omitempty" protobuf:"bytes,10,opt,name=keywords"`
+    
+    // Links is a map of human readable descriptions to URLs containing data about the application. This field can be 
+    // used to include URLs for licenses, dashboards, documentation, or other data pertaining to the application.
+    Links map[string] string `json:"links,omitempty" protobuf:"bytes,11,opt,name=links"`
```

Comment:
Some sort of "name" or "key" field will also be useful for programatically identifying and updating the links. For example, if one of the links is an elected master, it might be updated over time by an Operator.